### PR TITLE
Fix executor monitor bug

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
fix missing monitor callbacks 
#5644 
#5691

### CODE REVIEW

**Comment by piiswrong** [comment]
what was the problem?

File: src/executor/graph_executor.cc [hunk 1]
```diff
@@ -791,6 +811,12 @@ void GraphExecutor::RunOps(bool is_train, size_t topo_start, size_t topo_end) {
       bool profiling = false;
 #endif
       Engine::Get()->Push(seg_op.opr, seg_op.ctx, 0, profiling);
+      // Monitor callback for the segment
+      if (monitor_callback_) {
```
Comment (piiswrong) [inline src/executor/graph_executor.cc]: bulk should be disabled when using monitor

**Comment by eric-haibin-lin** [comment]
I think the problem is that, by the time the executor is initialized, it doesn't have a monitor callback installed. If the monitor is installed later and runs forward(), the executor still performs the bulk execution. The pr doesn't fix the issue, I'll work on this later.

File: src/executor/graph_executor.cc [hunk 2]
```diff
@@ -779,8 +801,6 @@ void GraphExecutor::RunOps(bool is_train, size_t topo_start, size_t topo_end) {
   }
 
   // Push Ops
-  static const auto& flist_outputs =
-      nnvm::Op::GetAttr<nnvm::FListOutputNames>("FListOutputNames");
   for (size_t nid = topo_start; nid < topo_end; ++nid) {
     auto seg_op = cached_seg_opr_[nid];
     // Check segments first
```
Comment (piiswrong) [inline src/executor/graph_executor.cc]: I think you can skip bulk here by checking monitor_callback_

File: src/executor/graph_executor.cc [hunk 3]
```diff
@@ -767,6 +767,28 @@ void GraphExecutor::InitOpSegs() {
   return;
 }
 
+void GraphExecutor::ExecuteMonCallback(size_t nid) {
+  static const auto& flist_outputs =
+      nnvm::Op::GetAttr<nnvm::FListOutputNames>("FListOutputNames");
+  const auto& idx = graph_.indexed_graph();
```
Comment (piiswrong) [inline src/executor/graph_executor.cc]: This might be slow. Is it doing topo sort every time?
How about input const OpNode& instead

Comment (eric-haibin-lin) [inline src/executor/graph_executor.cc]: https://github.com/dmlc/nnvm/blob/0aa3ff6fd561789b27242f257593962080f2de88/src/core/graph.cc#L12
It's not constructing a new graph every time.

Comment (piiswrong) [inline src/executor/graph_executor.cc]: i see

**Pull request merged**